File: cpp/cell.cc

#include <map>
#include <vector>

class Cell {
    
    //should make this private but too lazy to make a getter/setter for each
    public:
        int row, col;
        Cell* north;
        Cell* south;
        Cell* east;
        Cell* west;
        std::map<Cell*,bool> links;
    //public:

        Cell(int row_, int col_) {
            row = row_;
            col = col_;
        }
        
        void link(Cell* cell, bool bidi=true){
            links[cell] = true;
            if (bidi)
                cell -> link(this, false);
        }
        void unlink(Cell* cell, bool bidi = true) {
            links.erase(cell);
            if (bidi)
                cell -> unlink(this, false);
        }
        std::vector<Cell*> get_links(){
            std::vector<Cell*> vec;
            for(auto const& val : links)
                vec.push_back(val.first);
            return vec;
        }
        bool is_linked(Cell* cell){
            return links.contains(cell);
        }
        std:: vector<Cell*> get_neighbors(){
            std::vector<Cell*> vec;

            if (north != nullptr)
                vec.push_back(north);
            if(south != nullptr)
                vec.push_back(south);
            if(east != nullptr)
                vec.push_back(east);
            if(west != nullptr)
                vec.push_back(west);
            return vec;

        }
};

File: python/cell.py
from distances import Distances

class Cell:
    def __init__(self, row, col):
        self.row = row
        self.col = col
        self.north = None
        self.south = None
        self.east = None
        self.west = None
        self.links = {}
        #self.distances = None

    def link(self, cell, bidi=True):
        self.links[cell] = True
        if bidi:
            cell.link(self, False)

    def unlink(self, cell, bidi=True):
        self.links.pop(cell)
        if bidi:
            cell.unlink(self, False)

    def get_links(self):
        return list(self.links.keys())

    def is_linked(self, cell):
        return cell in self.links

    def get_neighbors(self):
        neighbors = []
        if self.north:
            neighbors.append(self.north)
        if self.south:
            neighbors.append(self.south)
        if self.east:
            neighbors.append(self.east)
        if self.west:
            neighbors.append(self.west)
        return neighbors
    
    def distances(self):
        distances = Distances(self)
        frontier = [self]
        while len(frontier) > 0:
            new_frontier = []
            for cell in frontier: 
                for linked in cell.get_links():
                    #! Might not work. Possible replacement: if self.distances[linked]
                    if linked in distances: 
                        continue
                    distances[linked] = distances[cell] + 1
                    new_frontier.append(linked)
            frontier = new_frontier
        return distances